prompttools/experiment/experiments/langchain_experiment.py

```diff
@@ -0,0 +1,128 @@
+        self.prompt = kwargs["prompt"]
+        self.call_params = {k: kwargs[k] for k in self.CALL_PARAMETERS if k in self.CALL_PARAMETERS}
+        self.output_keys = kwargs["output_key"]
+        self.llms = []
+        # Only supporting temperature for now
+        for llm in llms:
+            for temp in kwargs["temperature"]:
+                self.llms.append(llm(temperature=temp))
+        if os.getenv("DEBUG", default=False):
+            self.completion_fn = mock_lc_completion_fn
+        self.chain_params = list(zip(kwargs["prompt_chain"], kwargs["output_key"], strict=True))
+        self.all_args = {
+            "temperature": kwargs["temperature"],
+            "llms": llms,
+            "prompt": self.prompt
+        }
+        super().__init__()
+
+    def prepare_chain(self) -> None:
+        """
+        Links each individual chain to assemble
+        a sequential chain.
+        """
+        self.seq_chains = []
+        for llm in self.llms:
+            seq_chain = [
+                LLMChain(llm=llm, prompt=prompt, output_key=output_key)
+                for prompt, output_key in self.chain_params
+                ]
+            overall_chain = SequentialChain(
+                chains=seq_chain,
+                input_variables=self.call_params["input_variables"],
+                output_variables=self.call_params["output_variables"],
+                verbose=True,
+            )
+            self.seq_chains.append(overall_chain)
+
+    def lc_completion_fn(
+        self,
+        seq_chain,
+    ):
+        r"""
+        Makes request to LLM using the assembled
+        sequential chain.
+        """
+        return seq_chain(self.prompt)
+
+    def run(
+        self,
+        runs: int = 1,
+    ) -> None:
+        r"""
+        Create tuples of input and output for every possible combination of arguments.
+        For each combination, it will execute `runs` times, default to 1.
+        # TODO This can be done with an async queue
+        """
+        if not self.argument_combos:
+            logging.info("Preparing first...")
+            self.prepare_chain()
+        self.results = []
+        for seq_chain in self.seq_chains:
+            for _ in range(runs):
+                start = perf_counter()
+                res = self.completion_fn(seq_chain)
+                self.scores["latency"].append(perf_counter() - start)
+                self.results.append(res)
+                self.argument_combos.append({
+                    "llm": seq_chain.chains[0].llm,
+                    "temperature": seq_chain.chains[0].llm.temperature,
+                    "prompt": self.prompt,
+                    })
+        if len(self.results) == 0:
+            logging.error("No results. Something went wrong.")
+            raise PromptExperimentException
+
+    @staticmethod
+    def _extract_responses(output: List[Dict[str, object]]) -> list[str]:
+        return [str(output)]
```

Comment:
Small tweak: We should return a string here. I need to clean up the type annotations in other files to return a string instead of a list.

Reply:
Maybe this calls for a template we can use across all experiments?